pkg/openstack/common.go

```diff
@@ -815,3 +821,73 @@ func DeleteCertificate(
 	helper.GetLogger().Info(fmt.Sprintf("Deleting cert %s", certName))
 	return cert.Delete(ctx, helper)
 }
+
+func DeleteCertsAndRoutes(
+	ctx context.Context,
+	instance *corev1beta1.OpenStackControlPlane,
+	helper *helper.Helper,
+) (ctrl.Result, error) {
+
+	log := GetLogger(ctx)
+
+	// Retrieve all routes, certs and services in the namespace
+	routes, err := GetRoutesListWithLabel(ctx, helper, instance.Namespace, nil)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get routes: %w", err)
+	}
+
+	certs := &certmgrv1.CertificateList{}
+	if err := helper.GetClient().List(ctx, certs, client.InNamespace(instance.Namespace)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get certificates: %w", err)
+	}
+
+	services := &k8s_corev1.ServiceList{}
+	if err := helper.GetClient().List(ctx, services, client.InNamespace(instance.Namespace)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get services: %w", err)
+	}
+
+	var delErrs []error
+	for _, route := range routes.Items {
+		serviceExists := false
+		for _, svc := range services.Items {
+			if svc.Name == route.Spec.To.Name {
+				serviceExists = true
+				break
+			}
+		}
+
+		if serviceExists {
+			continue
+		}
+
+		if !object.CheckOwnerRefExist(instance.GetUID(), route.OwnerReferences) {
+			continue
+		}
+
+		// Delete related certs by route-basename and owner
+		for _, cert := range certs.Items {
+			routeBaseName := strings.TrimSuffix(route.Name, "-public")
```

Comment:
this should be outside of the certs.Items loop as it does not change during the loop so it should not be re-calculated each time